File: IntelPresentMon/AppWinUI/Services/StartupOptions.cs
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: MIT
using System;
using System.IO;

namespace PresentMon.UI.Services;

public sealed record StartupOptions
{
    public const string DefaultMutexSuffix = "UiBrowserProcess";

    public string PipeName { get; init; } = string.Empty;
    public bool FilesWorking { get; init; }
    public string MutexSuffix { get; init; } = DefaultMutexSuffix;
    public string DataDirectory { get; init; } = string.Empty;
    public string AppDataDirectory { get; init; } = string.Empty;
    public string InstallDirectory { get; init; } = string.Empty;
    public string LogDirectory { get; init; } = string.Empty;
    public bool EnableDevOptions { get; init; }

    public static StartupOptions Parse(string[] args)
    {
        ArgumentNullException.ThrowIfNull(args);
        var pipeName = string.Empty;
        var filesWorking = false;
        var mutexSuffix = DefaultMutexSuffix;
        var logDirectory = string.Empty;
        var dataDirectory = string.Empty;
        var enableDevOptions = false;

        for (var index = 0; index < args.Length; index++)
        {
            var argument = args[index];
            var separator = argument.IndexOf('=');
            var name = separator < 0 ? argument : argument[..separator];
            string ReadValue()
            {
                var value = separator >= 0 ? argument[(separator + 1)..]
                    : index + 1 < args.Length && !args[index + 1].StartsWith("--", StringComparison.Ordinal)
                        ? args[++index] : string.Empty;
                if (string.IsNullOrWhiteSpace(value))
                {
                    throw new ArgumentException($"The {name} option requires a value.", nameof(args));
                }
                return value;
            }

            bool ReadFlag()
            {
                if (separator < 0)
                {
                    return true;
                }
                return bool.TryParse(argument[(separator + 1)..], out var value)
                    ? value : throw new ArgumentException($"The {name} option requires true or false.", nameof(args));
            }

            switch (name)
            {
                case "--p2c-act-name":
                    pipeName = ReadValue();
                    break;
                case "--p2c-files-working":
                    filesWorking = ReadFlag();
                    break;
                case "--p2c-ui-mutex-name":
                    mutexSuffix = ReadValue();
                    break;
                case "--p2c-log-folder":
                    logDirectory = ReadValue();
                    break;
                case "--p2c-enable-ui-dev-options":
                    enableDevOptions = ReadFlag();
                    break;
                case "--data-directory":
                    dataDirectory = ReadValue();
                    break;
                // The kernel also forwards legacy CEF and logging switches.
                // Unrecognized arguments are intentionally ignored, as in CEF.
            }
        }

        string appDataDirectory;
        if (!string.IsNullOrEmpty(dataDirectory))
        {
            dataDirectory = Path.GetFullPath(dataDirectory);
            appDataDirectory = dataDirectory;
        }
        else if (filesWorking)
        {
            dataDirectory = Environment.CurrentDirectory;
            appDataDirectory = dataDirectory;
        }
        else
        {
            dataDirectory = Path.Combine(GetKnownFolder(Environment.SpecialFolder.MyDocuments), "PresentMon");
            appDataDirectory = Path.Combine(GetKnownFolder(Environment.SpecialFolder.LocalApplicationData), "Intel", "PresentMon");
        }

        var executableDirectory = Path.TrimEndingDirectorySeparator(Path.GetFullPath(AppContext.BaseDirectory));
        var installDirectory = string.Equals(Path.GetFileName(executableDirectory), "ui", StringComparison.OrdinalIgnoreCase)
            ? Directory.GetParent(executableDirectory)!.FullName : executableDirectory;
        return new StartupOptions
        {
            PipeName = pipeName,
            FilesWorking = filesWorking,
            MutexSuffix = mutexSuffix,
            DataDirectory = dataDirectory,
            AppDataDirectory = appDataDirectory,
            InstallDirectory = installDirectory,
            LogDirectory = string.IsNullOrWhiteSpace(logDirectory)
                ? Path.Combine(appDataDirectory, "logs") : Path.GetFullPath(logDirectory),
            EnableDevOptions = enableDevOptions,
        };
    }

    private static string GetKnownFolder(Environment.SpecialFolder folder)
    {
        var path = Environment.GetFolderPath(folder, Environment.SpecialFolderOption.DoNotVerify);
        if (string.IsNullOrWhiteSpace(path))
        {
            throw new DirectoryNotFoundException($"Windows did not provide the {folder} folder.");
        }
        return path;
    }
}

File: IntelPresentMon/AppWinUICpp/Core/StartupOptions.cpp
// Copyright (C) 2026 Intel Corporation
// SPDX-License-Identifier: MIT
#include "StartupOptions.h"

#include <Windows.h>
#include <ShlObj.h>

#include <filesystem>
#include <cctype>
#include <algorithm>
#include <stdexcept>
#include <string_view>

namespace pmon::ui::services
{
    namespace
    {
        std::string ToUtf8(const std::wstring& value)
        {
            if (value.empty()) {
                return {};
            }
            const auto size = WideCharToMultiByte(CP_UTF8, 0, value.data(), (int)value.size(), nullptr, 0, nullptr, nullptr);
            std::string output((size_t)size, '\0');
            WideCharToMultiByte(CP_UTF8, 0, value.data(), (int)value.size(), output.data(), size, nullptr, nullptr);
            return output;
        }

        std::filesystem::path ExecutableDirectory()
        {
            std::wstring path(MAX_PATH, L'\0');
            const auto length = GetModuleFileNameW(nullptr, path.data(), (DWORD)path.size());
            if (length == 0 || length == path.size()) {
                throw std::runtime_error("Unable to determine executable directory.");
            }
            path.resize(length);
            return std::filesystem::path{ path }.parent_path();
        }

        std::filesystem::path KnownFolder(REFKNOWNFOLDERID id)
        {
            PWSTR path = nullptr;
            const auto result = SHGetKnownFolderPath(id, KF_FLAG_DONT_VERIFY, nullptr, &path);
            if (FAILED(result) || path == nullptr) {
                throw std::runtime_error("Windows did not provide the requested known folder.");
            }
            const std::filesystem::path output{ path };
            CoTaskMemFree(path);
            return output;
        }

        std::string Absolute(const std::string& value)
        {
            return ToUtf8(std::filesystem::absolute(std::filesystem::u8path(value)).lexically_normal().wstring());
        }

        bool StartsWithOption(const std::string& argument)
        {
            return argument.rfind("--", 0) == 0;
        }

        std::string Lowercase(std::string value)
        {
            std::transform(value.begin(), value.end(), value.begin(), [](unsigned char character) {
                return (char)std::tolower(character);
            });
            return value;
        }
    }

    StartupOptions StartupOptions::Parse(const std::vector<std::string>& args)
    {
        StartupOptions options;
        std::string dataDirectory;
        std::string logDirectory;

        for (size_t index = 0; index < args.size(); ++index) {
            const auto& argument = args[index];
            const auto separator = argument.find('=');
            const auto name = argument.substr(0, separator);
            const auto readValue = [&]() {
                std::string value;
                if (separator != std::string::npos) {
                    value = argument.substr(separator + 1);
                }
                else if (index + 1 < args.size() && !StartsWithOption(args[index + 1])) {
                    value = args[++index];
                }
                if (value.empty()) {
                    throw std::invalid_argument("The " + name + " option requires a value.");
                }
                return value;
            };
            const auto readFlag = [&]() {
                if (separator == std::string::npos) {
                    return true;
                }
                const auto value = Lowercase(argument.substr(separator + 1));
                if (value == "true") {
                    return true;
                }
                if (value == "false") {
                    return false;
                }
                throw std::invalid_argument("The " + name + " option requires true or false.");
            };

            if (name == "--p2c-act-name") {
                options.PipeName = readValue();
            }
            else if (name == "--p2c-files-working") {
                options.FilesWorking = readFlag();
            }
            else if (name == "--p2c-ui-mutex-name") {
                options.MutexSuffix = readValue();
            }
            else if (name == "--p2c-log-folder") {
                logDirectory = readValue();
            }
            else if (name == "--p2c-enable-ui-dev-options") {
                options.EnableDevOptions = readFlag();
            }
            else if (name == "--data-directory") {
                dataDirectory = readValue();
            }
            else {
                throw std::invalid_argument("The " + name + " option is not supported by the WinUI interface.");
            }
        }

        if (!dataDirectory.empty()) {
            options.DataDirectory = Absolute(dataDirectory);
            options.AppDataDirectory = options.DataDirectory;
        }
        else if (options.FilesWorking) {
            options.DataDirectory = ToUtf8(std::filesystem::current_path().wstring());
            options.AppDataDirectory = options.DataDirectory;
        }
        else {
            options.DataDirectory = ToUtf8((KnownFolder(FOLDERID_Documents) / L"PresentMon").wstring());
            options.AppDataDirectory = ToUtf8((KnownFolder(FOLDERID_LocalAppData) / L"Intel" / L"PresentMon").wstring());
        }
        const auto executableDirectory = ExecutableDirectory();
        options.InstallDirectory = ToUtf8((Lowercase(ToUtf8(executableDirectory.filename().wstring())) == "ui"
            ? executableDirectory.parent_path() : executableDirectory).wstring());
        options.LogDirectory = logDirectory.empty()
            ? ToUtf8((std::filesystem::u8path(options.AppDataDirectory) / "logs").wstring()) : Absolute(logDirectory);
        return options;
    }
}
